Support the identification of the source host of the metric data in otel-receiver-plugin

wankai123 opened the pull request:
<!--
    ⚠️ Please make sure to read this template first, pull requests that don't accord with this template
    maybe closed without notice.
    Texts surrounded by `<` and `>` are meant to be replaced by you, e.g. <framework name>, <issue number>.
    Put an `x` in the `[ ]` to mark the item as CHECKED. `[x]`
-->

<!-- ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👇 ====
### Fix <bug description or the bug issue number or bug issue link>
- [ ] Add a unit test to verify that the fix works.
- [ ] Explain briefly why the bug exists and how to fix it.
     ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👆 ==== -->

<!-- ==== 🔌 Remove this line WHEN AND ONLY WHEN you're adding a new plugin, follow the checklist 👇 ====
### Add an agent plugin to support <framework name>
- [ ] Add a test case for the new plugin, refer to [the doc](https://github.com/apache/skywalking/blob/master/docs/en/guides/Plugin-test.md)
- [ ] Add a component id in [the component-libraries.yml](https://github.com/apache/skywalking/blob/master/oap-server/server-bootstrap/src/main/resources/component-libraries.yml)
- [ ] Add a logo in [the UI repo](https://github.com/apache/skywalking-rocketbot-ui/tree/master/src/views/components/topology/assets)
     ==== 🔌 Remove this line WHEN AND ONLY WHEN you're adding a new plugin, follow the checklist 👆 ==== -->

<!-- ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👇 ====
### Improve the performance of <class or module or ...>
- [ ] Add a benchmark for the improvement, refer to [the existing ones](https://github.com/apache/skywalking/blob/master/apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/LinkedArrayBenchmark.java)
- [ ] The benchmark result.
```text
<Paste the benchmark results here>
```
- [ ] Links/URLs to the theory proof or discussion articles/blogs. <links/URLs here>
     ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👆 ==== -->


### Add node information from the Openensus proto to the labels of the samples, to support the identification of the source host of the metric data
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [x] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.


- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

When I used the otel-receiver-plugin to receive prometheus node_exporter data forwarded from the otel-collector, 
I found that there is no label of the data source host in the samples, so I can't use MAL to do aggregation analysis by host,
unless I configure additional labels in otel-collector config.yaml, But then the label will be on each metrics.

In <opencensus.proto.agent.common.v1.Node>, node information is already carried, included  the host name.

So I add a labelkey 'node_identifier_host_name' to the collected data samples, the labelvalue is from 
<Node.identifier.host_name>, to support the identification of the source host of the metric data.

wu-sheng commented on docs/en/setup/backend/backend-receivers.md:
````diff
@@ -182,7 +182,9 @@ receiver-otel:
     enabledOcRules: ${SW_OTEL_RECEIVER_ENABLED_OC_RULES:"istio-controlplane"}
 ```
-
+The receiver will add a labelkey 'node_identifier_host_name' to the collected data samples，
+the labelvalue is from opencensus.proto.agent.common.v1.Node.identifier.host_name, 
+to support the identification of the source of the metric data.
````
```suggestion
The receiver adds a label with `key=node_identifier_host_name` to the collected data samples，
and value from `Node.identifier.host_name` defined in opencensus agent proto, 
to be the identification of the metric data.
```

wankai123 replied:
done

wu-sheng commented on oap-server/server-receiver-plugin/otel-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/otel/oc/OCMetricHandler.java:
```diff
@@ -107,6 +111,15 @@
         return result;
     }
 
+    private static Map<String, String> buildLabelsWithNodeInfo(Node node,
```
```suggestion
    private static Map<String, String> buildLabelsFromNodeInfo(Node node,
```

wu-sheng replied:
Please don't hide `buildLabels` in the `buildLabelsWithNodeInfo`, it is misleading the reader. Do like this `buildLabelsFromNodeInfo(buildLables(...))`.

wu-sheng commented on oap-server/server-receiver-plugin/otel-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/otel/oc/OCMetricHandler.java:
```diff
@@ -107,6 +111,15 @@
+    private static Map<String, String> buildLabelsWithNodeInfo(Node node,
+                                                               List<LabelKey> keys,
+                                                               List<LabelValue> values) {
+        Map<String, String> result = buildLabels(keys, values);
```
Take a look at this comment in the original proto
```
  // This is required only in the first message on the stream or if the
  // previous sent ExportMetricsServiceRequest message has a different Node (e.g.
  // when the same RPC is used to send Metrics from multiple Applications).
  opencensus.proto.agent.common.v1.Node node = 1;
```

You need to hold the previous node, if the current `request#hasNode()` is false, use the previous one.

wu-sheng commented:
Next time, please use a pure new branch to send the pull request. Otherwise, it always brings the previous commit log.

wankai123 commented:
done

wu-sheng commented on oap-server/server-receiver-plugin/otel-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/otel/oc/OCMetricHandler.java:
```diff
@@ -57,11 +58,21 @@
+
+            @Override
+            public void onNext(ExportMetricsServiceRequest request) {
+                if (request.hasNode())
+                    node = request.getNode();
```
I am not sure whether without `{}` is allowed. We will see.

hanahmily replied:
In case of NPE, I prefer to assign a default value to the `Node`, even node is sent in the first message.

wu-sheng replied:
Do you mean build an node with mock values of fields we need?

hanahmily replied:
Exactly, `Optional.orElse()` is what I like. Some defensive programming style.

hanahmily commented on oap-server/server-receiver-plugin/otel-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/otel/oc/OCMetricHandler.java:
```diff
@@ -107,6 +118,13 @@
+    private static Map<String, String> buildLabelsFromNodeInfo(Node node,
+                                                               Map<String, String> buildLabelsResult) {
+        if (node != null)
+            buildLabelsResult.put("node_identifier_host_name", node.getIdentifier().getHostName());
```
Could we add other fields more than `hostname` in `ProcessIdenifier`?

wu-sheng replied:
```
// Identifier that uniquely identifies a process within a VM/container.
message ProcessIdentifier {

  // The host name. Usually refers to the machine/container name.
  // For example: os.Hostname() in Go, socket.gethostname() in Python.
  string host_name = 1;

  // Process id.
  uint32 pid = 2;

  // Start time of this ProcessIdentifier. Represented in epoch time.
  google.protobuf.Timestamp start_timestamp = 3;
}
```

I think there is only `pid` makes sense? Do you think `_` is good, or should consider `.`? Such as `node.identifier.host_name`.

hanahmily replied:
MAL has a capacity to transform `.` to `_`, which means no matter what we choose, the final result must be `node_identifier_host_name `.

wu-sheng replied:
OK, then `_` is fine.

hanahmily replied:
A common use case about the `start_timestamp` is to get `uptime` metric, the expression should leverage `start_timestamp` and `MAL` 's `now()` function as below:

`now()-start_timestamp`

wu-sheng replied:
I am just wondering whether this is the meaning, as it is reported every time, rather than a separated metric.

wankai123 commented:
I added some defensive codes, and added a label with `key = node_identifier_pid` and value from `Node.identifier.pid` defined in opencensus agent proto.
I  don't add `start_timestamp` to the labels yet,  I think the `start_timestamp`  is a value and in MAL ladel is used for filter or aggregation, not for count.

wu-sheng reviewed (approved):
Generally good to me. @hanahmily Recheck please.

hanahmily reviewed (approved):
LGTM
